### PR TITLE
Use arc4rand(9) on FreeBSD

### PR DESCRIPTION
**Pull request by habnabit**
From rust-lang-nursery/rand#112:

> After reading through #30691 it seems that there's general agreement that using OS-provided facilities for seeding rust userland processes is fine as long as it doesn't use too much from libc. FreeBSD's `arc4random_buf(3)` is not only a whole lot of libc code, but also not even currently exposed in the libc crate. Fortunately, the mechanism `arc4random_buf(3)` et al. use for getting entropy from the kernel ([`arc4rand(9)`](https://www.freebsd.org/cgi/man.cgi?query=arc4random&apropos=0&sektion=9&manpath=FreeBSD+10.3-RELEASE&arch=default&format=html)) is exposed via `sysctl(3)` with constants that are already in the libc crate.
> 
> I haven't found too much documentation on `KERN_ARND`—it's missing or only briefly described in most of the places that cover sysctl mibs. But, from digging through the kernel source, it appears that the sysctl used in this PR is very close to just calling `arc4rand(9)` directly (with `reseed` set to 0 and no way to change it).

I expected [rand](/rust-lang-nursery/rand) to reply quicker, so I tried submitting it there first. It's been a few weeks with no comment, so I don't know the state of it, but maybe someone will see it here and have an opinion. This is basically the same patch. It pains me to duplicate the code but I guess it hasn't been factored out into just one place yet.

### CODE REVIEW

File: src/libstd/sys/unix/rand.rs [hunk 1]
```diff
@@ -10,14 +10,31 @@
...
+    let mut buf: [u8; 4] = [0; 4];
+    fill_buf(&mut buf);
+    unsafe { mem::transmute::<[u8; 4], u32>(buf) }
+}
+
+fn next_u64(mut fill_buf: &mut FnMut(&mut [u8])) -> u64 {
+    let mut buf: [u8; 8] = [0; 8];
+    fill_buf(&mut buf);
+    unsafe { mem::transmute::<[u8; 8], u64>(buf) }
+}
```
Comment (eddyb) [inline src/libstd/sys/unix/rand.rs]: Why are these not using a generic `F: FnMut(&mut [u8])`?

Comment (habnabit) [inline src/libstd/sys/unix/rand.rs]: No particular reason, I suppose.

**Comment by brson** [comment]
Sorry nobody saw your rand patch.

Conceptually this seems good to me. Haven't read the patch closely though.

The rand code in-tree and in the rand crate have effectively diverged afaik. The in-tree crate is just for std hashmaps and whatever rustc needs, and makes some different tradeoffs.

**Comment by alexcrichton** [comment]
Discussed during libs triage today the conclusion was to merge, thanks for the patch @habnabit!

@bors: r+

**Pull request merged**